src/main/java/org/spongepowered/api/item/inventory/menu/InventoryMenu.java

```diff
@@ -0,0 +1,167 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory.menu;
+
+import org.spongepowered.api.entity.living.player.Player;
+import org.spongepowered.api.item.inventory.Container;
+import org.spongepowered.api.item.inventory.custom.ContainerType;
+import org.spongepowered.api.item.inventory.slot.SlotIndex;
+import org.spongepowered.api.item.inventory.type.ViewableInventory;
+import org.spongepowered.api.text.Text;
+
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+/**
+ * Helper for Menus based on Inventories.
+ * <p>This helper provides simple callbacks that can be used instead of listening to inventory events.</p>
+ * <p>InventoryMenus are by default readonly and automatically prevent any changes made by players in the menu.</p>
+ */
+public interface InventoryMenu {
+
+    /**
+     * Creates a new InventoryMenu based on given inventory.
+     *
+     * @param inventory the inventory
+     *
+     * @return the new menu.
+     */
+    static InventoryMenu of(ViewableInventory inventory) {
+        return inventory.asMenu();
+    }
+
+    /**
+     * Returns the current inventory used in this menu.
+     *
+     * @return the current inventory
+     */
+    ViewableInventory getCurrentInventory();
+
+    /**
+     * Returns the container type of the current inventory.
+     *
+     * @return current container type.
+     */
+    ContainerType getType();
+
+    /**
+     * Sets a new inventory. If the ContainerType does not change the inventory will be swapped out silently.
+     * <p>If the ContainerType is different all existing callbacks are cleared and open menus are closed and reopened with the new inventory.</p>
+     *
+     * @param inventory the new inventory
+     */
+    void setCurrentInventory(ViewableInventory inventory);
+
+    /**
+     * Sets the title of this menu.
+     * <p>Any open menus are closed and reopened with the new title.</p>
+     *
+     * @param title the new title.
+     */
+    void setTitle(Text title);
+
+    /**
+     * Registers click callbacks. Unregisters previously registered handler of the same type.
+     * <p>Return false in the handler to prevent changes.</p>
+     * <p>Possible handlers: {@link ClickHandler}, {@link SlotClickHandler}, {@link KeySwapHandler}, {@link SlotChangeHandler}, {@link CloseHandler}</p>
+     *
+     * @param handler the handler
+     */
+    void registerHandler(Object handler);
```

Comment:
Y u no union type java :/

Reply:
I agree with a marker interface here. Otherwise this encourages someone to call whatever to it.

It would be different if you made the handlers be annotated methods.